connectors/services/job_executor.py

```diff
@@ -0,0 +1,228 @@
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""
+Event loop
+
+- polls for work by calling Elasticsearch on a regular basis
+- instantiates connector plugins
+- mirrors an Elasticsearch index with a collection of documents
+"""
+import asyncio
+import os
+import time
+
+from connectors.byoc import BYOIndex, DataSourceError, PipelineSettings, SyncJobIndex
+from connectors.byoei import ElasticServer
+from connectors.es import defaults_for
+from connectors.logger import logger
+from connectors.services.base import BaseService
+from connectors.source import DataSourceConfiguration, get_source_klass
+
+
+class SyncJobRunner:
+    def __init__(
+        self,
+        sources,
+        sync_job,
+        elastic_server,
+        connector_index,
+        sync_job_index,
+        bulk_options,
+    ):
+        self.sources = sources
+        self.sync_job = sync_job
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self.connector_index = connector_index
+        self.sync_job_index = sync_job_index
+
+    async def execute(self):
+        await self.sync_job.claim()
+
+        start_time = time.time()
+        index_name = self.sync_job.connector["index_name"]
+        service_type = self.sync_job.connector["service_type"]
+        language_code = self.sync_job.connector["language"]
+        pipeline = PipelineSettings(self.sync_job.connector.get("pipeline", {}))
+
+        fqn = self.sources[service_type]
+        configuration = DataSourceConfiguration(
+            self.sync_job.connector["configuration"]
+        )
+        try:
+            source_klass = get_source_klass(fqn)
+        except Exception as e:
+            logger.critical(e, exc_info=True)
+            raise DataSourceError(f"Could not instantiate {fqn} for {service_type}")
+
+        data_provider = source_klass(configuration)
+
+        if not await data_provider.changed():
+            logger.debug(f"No change in {service_type} data provider, skipping...")
+            return
+
+        logger.debug(f"Syncing '{service_type}'")
+        try:
+            logger.debug(f"Pinging the {source_klass} backend")
+            await data_provider.ping()
+            await asyncio.sleep(0)
+
+            mappings, settings = defaults_for(
+                is_connectors_index=True, language_code=language_code
+            )
+            logger.debug("Preparing the index")
+            await self.elastic_server.prepare_index(
+                index_name, mappings=mappings, settings=settings
+            )
+            await asyncio.sleep(0)
+
+            result = await self.elastic_server.async_bulk(
+                index_name,
+                self.prepare_docs(pipeline, data_provider),
+                pipeline,
+                options=self.bulk_options,
+            )
+            for i in range(50):
```

Comment:
That block needs to be removed. or replaced by

```
await asyncio.sleep(5)
```

when you call sleep, you have already sent back the execution to the event loop, no need to loop on smaller sleeps